File: scripts/archive/import_quizzes.py
import firebase_admin
from firebase_admin import credentials, firestore
import re
import time
import sys

SERVICE_ACCOUNT_PATH = r"C:\Users\xioas\.gemini\antigravity\scratch\msdl\serviceAccountKey.json"
INPUT_PATH = r"C:\Users\xioas\.gemini\antigravity\scratch\msdl\all_user_input.txt"

def parse_quizzes(text):
    quizzes = []
    
    # Split by "Question 1", "Question 2", etc.
    # Actually, split by "\nQuestion\n"
    # But wait, in the data, it's "Question 1\n\nQuestion\nNajasat..."
    
    parts = re.split(r'\nQuestion\n', text)
    
    for i, part in enumerate(parts):
        if i == 0:
            continue
        
        try:
            q_match = re.search(r'^(.*?)\n+Option 1\n', part, re.DOTALL)
            o1_match = re.search(r'Option 1\n(.*?)\n+Option 2\n', part, re.DOTALL)
            o2_match = re.search(r'Option 2\n(.*?)\n+Option 3\n', part, re.DOTALL)
            o3_match = re.search(r'Option 3\n(.*?)\n+Option 4\n', part, re.DOTALL)
            o4_match = re.search(r'Option 4\n(.*?)\n+Correct Answer\n', part, re.DOTALL)
            ca_match = re.search(r'Correct Answer\n(.*?)\n+Category\n', part, re.DOTALL)
            
            if not (q_match and o1_match and o2_match and o3_match and o4_match and ca_match):
                continue
                
            question = q_match.group(1).strip()
            o1 = o1_match.group(1).strip()
            o2 = o2_match.group(1).strip()
            o3 = o3_match.group(1).strip()
            o4 = o4_match.group(1).strip()
            ca = ca_match.group(1).strip()
            
            cat_match = re.search(r'Category\n([^\n]+)', part)
            if not cat_match:
                continue
            cat = cat_match.group(1).strip()
            
            quiz = {
                "question": question,
                "option1": o1,
                "option2": o2,
                "option3": o3,
                "option4": o4,
                "correctAnswer": ca,
                "category": cat
            }
            quizzes.append(quiz)
        except Exception as e:
            continue
            
    return quizzes

def validate_quiz(quiz):
    required = ["question", "option1", "option2", "option3", "option4", "correctAnswer", "category"]
    for r in required:
        if r not in quiz or not isinstance(quiz[r], str) or quiz[r] == "":
            return False, f"Missing or empty field: {r}"
            
    ca = quiz["correctAnswer"]
    options = [quiz["option1"], quiz["option2"], quiz["option3"], quiz["option4"]]
    if ca not in options:
        return False, f"correctAnswer '{ca}' does not match any option"
        
    return True, ""

def main():
    try:
        cred = credentials.Certificate(SERVICE_ACCOUNT_PATH)
        firebase_admin.initialize_app(cred)
        print("Authentication: SUCCESS")
    except Exception as e:
        print(f"Authentication: FAILED\n{e}")
        return
        
    try:
        db = firestore.client()
        print("Firestore: CONNECTED")
    except Exception as e:
        print(f"Firestore: FAILED\n{e}")
        return
    
    with open(INPUT_PATH, 'r', encoding='utf-8') as f:
        text = f.read()
        
    parsed_quizzes = parse_quizzes(text)
    
    existing_questions = set()
    try:
        docs = db.collection('quizzes').stream()
        for doc in docs:
            data = doc.to_dict()
            if 'question' in data:
                existing_questions.add(data['question'])
    except Exception as e:
        print(f"Firestore: FAILED\n{e}")
        return
        
    start_time = time.time()
    
    inserted_count = 0
    skipped_count = 0
    validation_errors = 0
    failed_writes = 0
    
    category_counts = {}
    valid_new_quizzes = []
    
    for q in parsed_quizzes:
        is_valid, err = validate_quiz(q)
        if not is_valid:
            validation_errors += 1
            continue
            
        if q["question"] in existing_questions:
            skipped_count += 1
            continue
            
        valid_new_quizzes.append(q)
        existing_questions.add(q["question"])
        
    batch_count = 0
    batch_size = 400
    
    for i in range(0, len(valid_new_quizzes), batch_size):
        chunk = valid_new_quizzes[i:i+batch_size]
        batch = db.batch()
        
        for q in chunk:
            doc_ref = db.collection('quizzes').document()
            batch.set(doc_ref, q)
            
        try:
            batch.commit()
            batch_count += 1
            inserted_count += len(chunk)
            
            for q in chunk:
                cat = q['category']
                category_counts[cat] = category_counts.get(cat, 0) + 1
                
        except Exception as e:
            failed_writes += len(chunk)
            print(f"Batch {batch_count + 1} failed: {e}")
            break
            
    total_quizzes = len(existing_questions)
    exec_time = time.time() - start_time
    
    print("\nDocuments Processed:")
    print(f"Inserted: {inserted_count}")
    print(f"Skipped (Duplicates): {skipped_count}")
    print(f"Validation Errors: {validation_errors}")
    print(f"Failed Writes: {failed_writes}")
    print(f"Batch Count: {batch_count}")
    
    print("\nCategory Counts:")
    for cat, count in category_counts.items():
        print(f"{cat}: {count}")
        
    print(f"\nFinal Total Quiz Count: {total_quizzes}")
    print(f"Execution Time: {exec_time:.2f} seconds")
    
    if failed_writes > 0:
        print("\nImport Status: FAILED")
    elif inserted_count > 0:
        print("\nImport Status: SUCCESS")
    else:
        print("\nImport Status: SUCCESS")

if __name__ == "__main__":
    main()

File: scripts/archive/import_quizzes.js

const { getFirestore } = require('firebase-admin/firestore');
const fs = require('fs');

const SERVICE_ACCOUNT_PATH = 'C:\\Users\\xioas\\.gemini\\antigravity\\scratch\\msdl\\serviceAccountKey.json';
const INPUT_PATH = 'C:\\Users\\xioas\\.gemini\\antigravity\\scratch\\msdl\\all_user_input.txt';

function parseQuizzes(text) {
  const quizzes = [];
  
  // A regex to match an entire quiz question block
  // We use [\s\S] to match across newlines.
  // The block looks like:
  // Question
  // <question_text>
  // Option 1
  // <opt1>
  // Option 2
  // <opt2>
  // Option 3
  // <opt3>
  // Option 4
  // <opt4>
  // Correct Answer
  // <ans>
  // Category
  // <cat>
  
  const blockRegex = /Question\s+([\s\S]*?)\s+Option 1\s+([\s\S]*?)\s+Option 2\s+([\s\S]*?)\s+Option 3\s+([\s\S]*?)\s+Option 4\s+([\s\S]*?)\s+Correct Answer\s+([\s\S]*?)\s+Category\s+([^\r\n]+)/g;
  
  let match;
  while ((match = blockRegex.exec(text)) !== null) {
    let question = match[1].trim();
    // Sometimes 'Question' matched the "Question 1\n\nQuestion\n..." part, so we need to clean up question text.
    // If question text starts with something like "1\n\nQuestion\n", strip it.
    question = question.replace(/^\d+\s+Question\s+/, '').trim();
    
    const quiz = {
      question: question,
      option1: match[2].trim(),
      option2: match[3].trim(),
      option3: match[4].trim(),
      option4: match[5].trim(),
      correctAnswer: match[6].trim(),
      category: match[7].trim()
    };
    quizzes.push(quiz);
  }
  return quizzes;
}

function validateQuiz(quiz) {
  const required = ["question", "option1", "option2", "option3", "option4", "correctAnswer", "category"];
  for (const r of required) {
    if (!quiz[r] || typeof quiz[r] !== 'string' || quiz[r].trim() === "") {
      return { isValid: false, err: `Missing or empty field: ${r}` };
    }
  }
  
  const ca = quiz.correctAnswer;
  const options = [quiz.option1, quiz.option2, quiz.option3, quiz.option4];
  if (!options.includes(ca)) {
    return { isValid: false, err: `correctAnswer '${ca}' does not match any option` };
  }
  
  return { isValid: true, err: "" };
}

async function main() {
  console.log("Authentication: PENDING");
  try {
    const serviceAccount = require(SERVICE_ACCOUNT_PATH);
    const { initializeApp, cert } = require('firebase-admin/app');
    initializeApp({
      credential: cert(serviceAccount)
    });
    console.log("Authentication: SUCCESS");
  } catch (e) {
    console.log(`Authentication: FAILED (${e.message})`);
    return;
  }
  
  console.log("Firestore: CONNECTED");
  const db = getFirestore();
  
  const text = fs.readFileSync(INPUT_PATH, 'utf-8');
  const parsedQuizzes = parseQuizzes(text);
  
  // Debug output
  console.log(`Parsed Quizzes Count: ${parsedQuizzes.length}`);
  
  const existingQuestions = new Set();
  try {
    const snapshot = await db.collection('quizzes').get();
    snapshot.forEach(doc => {
      const data = doc.data();
      if (data.question) {
        existingQuestions.add(data.question);
      }
    });
  } catch (e) {
    console.log(`Firestore: FAILED (${e.message})`);
    return;
  }
  
  const startTime = Date.now();
  
  let insertedCount = 0;
  let skippedCount = 0;
  let validationErrors = 0;
  let failedWrites = 0;
  
  const categoryCounts = {};
  const validNewQuizzes = [];
  
  for (const q of parsedQuizzes) {
    const { isValid, err } = validateQuiz(q);
    if (!isValid) {
      // console.log("Validation Error:", err, q);
      validationErrors++;
      continue;
    }
    
    if (existingQuestions.has(q.question)) {
      skippedCount++;
      continue;
    }
    
    validNewQuizzes.push(q);
    existingQuestions.add(q.question);
  }
  
  let batchCount = 0;
  const batchSize = 400; // Firebase limit is 500
  
  for (let i = 0; i < validNewQuizzes.length; i += batchSize) {
    const chunk = validNewQuizzes.slice(i, i + batchSize);
    const batch = db.batch();
    
    for (const q of chunk) {
      const docRef = db.collection('quizzes').doc();
      batch.set(docRef, q);
    }
    
    try {
      await batch.commit();
      batchCount++;
      insertedCount += chunk.length;
      
      for (const q of chunk) {
        const cat = q.category;
        categoryCounts[cat] = (categoryCounts[cat] || 0) + 1;
      }
    } catch (e) {
      failedWrites += chunk.length;
      console.log(`Batch ${batchCount + 1} failed: ${e.message}`);
      break; // Stop further writes
    }
  }
  
  const totalQuizzes = existingQuestions.size;
  const execTime = (Date.now() - startTime) / 1000;
  
  console.log("\nDocuments Processed:");
  console.log(`Inserted: ${insertedCount}`);
  console.log(`Skipped (Duplicates): ${skippedCount}`);
  console.log(`Validation Errors: ${validationErrors}`);
  console.log(`Failed Writes: ${failedWrites}`);
  console.log(`Batch Count: ${batchCount}`);
  
  console.log("\nCategory Counts:");
  for (const [cat, count] of Object.entries(categoryCounts)) {
    console.log(`${cat}: ${count}`);
  }
  
  console.log(`\nFinal Total Quiz Count: ${totalQuizzes}`);
  console.log(`Execution Time: ${execTime.toFixed(2)} seconds`);
  
  if (failedWrites > 0) {
    console.log("\nImport Status: FAILED");
  } else if (insertedCount > 0) {
    console.log("\nImport Status: SUCCESS");
  } else {
    console.log("\nImport Status: SUCCESS");
  }
}

main().catch(console.error);
